realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt

```diff
@@ -0,0 +1,1076 @@
+                    .sort(Document("hello", -1))
+            assertEquals(doc3, findOne(Document(), options).blockingGetResult()!!)
+        }
+    }
+
+    @Test
+    fun findOne_fails() {
+        with(getCollectionInternal()) {
+            assertFailsWith(ObjectServerError::class) {
+                findOne(Document("\$who", 1)).blockingGetResult()
+            }.also { e ->
+                assertEquals(ErrorCode.MONGODB_ERROR, e.errorCode)
+                assertNotNull(e.errorMessage)
+                assertTrue(e.errorMessage!!.contains("operator", true))
+            }
+        }
+    }
+
+    @Test
+    fun find() {
+        with(getCollectionInternal()) {
+            // Find on an empty collection returns false on hasNext and null on first
+            var iter = find()
+            assertFalse(iter.iterator().blockingGetResult()!!.hasNext())
+            assertNull(iter.first().blockingGetResult())
+
+            val doc1 = Document("hello", "world")
+            val doc2 = Document("hello", "friend")
+            doc2["proj"] = "field"
+            insertMany(listOf(doc1, doc2)).blockingGetResult()
+
+            // Iterate after inserting two documents
+            assertTrue(iter.iterator().blockingGetResult()!!.hasNext())
+            assertEquals(doc1, iter.first().blockingGetResult()!!.withoutId())
+
+            // Get next with sort by desc "_id" and limit to 1 document
+            assertEquals(doc2,
+                    iter.limit(1)
+                            .sort(Document("_id", -1))
+                            .iterator().blockingGetResult()!!
+                            .next().withoutId())
+
+            // Find first document
+            iter = find(doc1)
+            assertTrue(iter.iterator().blockingGetResult()!!.hasNext())
+            assertEquals(doc1,
+                    iter.iterator().blockingGetResult()!!
+                            .next().withoutId())
+
+            // Find with filter for first document
+            iter = find().filter(doc1)
+            assertTrue(iter.iterator().blockingGetResult()!!.hasNext())
+            assertEquals(doc1,
+                    iter.iterator().blockingGetResult()!!
+                            .next().withoutId())
+
+            // Find with projection shows "proj" in result
+            val expected = Document("proj", "field")
+            assertEquals(expected,
+                    find(doc2)
+                            .projection(Document("proj", 1))
+                            .iterator().blockingGetResult()!!
+                            .next().withoutId())
+
+            // Getting a new iterator returns first element on tryNext
+            val asyncIter = iter.iterator().blockingGetResult()!!
+            assertEquals(doc1, asyncIter.tryNext().withoutId())
+        }
+    }
+
+    @Test
+    fun find_fails() {
+        with(getCollectionInternal()) {
+            assertFailsWith(ObjectServerError::class) {
```

Comment:
There is already a `assertFailsWithErrorCode(expectedCode: ErrorCode, method: () -> Unit)` in `KotlinTestUtils.kt` maybe build on top of that. Could be changes in to a chain'able construct, so you could assert further details like you do.